org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -53,33 +56,134 @@
 		if (unit == null) {
 			return Collections.emptyList();
 		}
+
+		int start = DiagnosticsHelper.getStartOffset(unit, params.getRange());
+		int end = DiagnosticsHelper.getEndOffset(unit, params.getRange());
+		InnovationContext context = new InnovationContext(unit, start, end - start);
+		context.setASTRoot(getASTRoot(unit));
+		// TODO (Yan): Problem locations should be calculated only for quickfix code actions.
+		IProblemLocationCore[] locations = this.getProblemLocationCores(unit, params.getContext().getDiagnostics());
+
+		List<String> codeActionKinds = new ArrayList<>();
+		if (params.getContext().getOnly() != null && !params.getContext().getOnly().isEmpty()) {
+			codeActionKinds.addAll(params.getContext().getOnly());
+		} else {
+			List<String> defaultCodeActionKinds = Arrays.asList(
+				CodeActionKind.QuickFix,
+				CodeActionKind.Refactor,
+				CODE_ACTION_KIND_QUICK_ASSIST,
+				CodeActionKind.Source
+			);
+			codeActionKinds.addAll(defaultCodeActionKinds);
+		}
+
+		List<ChangeCorrectionProposal> proposals = new ArrayList<>();
+		if (codeActionKinds.contains(CodeActionKind.QuickFix)) {
+			try {
+
+				List<ChangeCorrectionProposal> quickfixProposals = this.quickFixProcessor.getCorrections(context, locations);
+				proposals.addAll(quickfixProposals);
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem resolving quick fix code actions", e);
+			}
+		}
+
+		if (codeActionKinds.contains(CodeActionKind.Refactor)) {
+			try {
+				List<ChangeCorrectionProposal> refactorProposals = this.refactorProcessor.getProposals(params, context);
+				proposals.addAll(refactorProposals);
+
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem resolving refactor code actions", e);
+			}
+		}
+
+		if (codeActionKinds.contains(CODE_ACTION_KIND_QUICK_ASSIST)
+			// TODO (Yan): many refactor actions are still in quickAssistProcessor now, should move them to refactorProcessor and remove below condition later
+			|| codeActionKinds.contains(CodeActionKind.Refactor)
+		) {
+			try {
+				List<ChangeCorrectionProposal> quickassistProposals = this.quickAssistProcessor.getAssists(params, context, locations);
+				proposals.addAll(quickassistProposals);
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem resolving quick assist code actions", e);
+			}
+			// TODO (Yan): Remove below block, move InvertBoolean proposals into RefactorProcessor
+			try {
+				List<CUCorrectionProposal> corrections = this.advancedQuickAssistProcessor.getAssists(params, context, locations);
+				proposals.addAll(corrections);
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem resolving advanced quick assist code actions", e);
+			}
+		}
+
+		// TODO (Yan): below comparator seldomly compares relavance, it returns early if code action kind is different.
+		proposals.sort(new ChangeCorrectionProposalComparator());
```

Comment:
The sorting issue is a new issue, please open an issue to track it instead of TODO tag.

Reply:
https://github.com/eclipse/eclipse.jdt.ls/issues/1250